### ISSUE TITLE
Liquibase VARCHAR limit with MariaDB/MySQL and UTF8

### ISSUE THREAD

**Issue by joshuaboniface** [Libresonic/libresonic#233]
First a big thanks to @muff1nman for the Liquibase support, for me it's been a long-anticipated feature!

I've been able to get the database working fine with Tomcat 8 and the MariaDB driver (using 10.1 on the database side) on the `develop` branch. However if I use UTF8 for my database, I run into trouble with the size of the `VARCHAR`s in many tables, specifically `user`:

> liquibase.exception.MigrationFailedException: Migration failed for change set classpath:liquibase/legacy/schema25.xml::schema25_004::muff1nman:
>      Reason: liquibase.exception.DatabaseException: (conn:515567) Index column size too large. The maximum column size is 767 bytes.
> Query is : CREATE TABLE libresonic_utf.user (username VARCHAR(512) NOT NULL, password VARCHAR(512) NOT NULL, CONSTRAINT PK_USER PRIMARY KEY (username)) [Failed SQL: CREATE TABLE libresonic_utf.user (username VARCHAR(512) NOT NULL, passwor$
>  VARCHAR(512) NOT NULL, CONSTRAINT PK_USER PRIMARY KEY (username))]

This is a common limitation of MySQL with UTF8 encoding, which I do prefer to use for its flexibility.

There are two obvious fixes:
1. Use another (smaller) character set for the database, such as `latin1`. (User-side)
2. Modify the Liquibase schema in `libresonic-main/src/main/resources/applicationContext-db.xml` to use a smaller 255 character length for `VARCHAR` columns. (Project-side)

I'm opening this issue mostly as a discussion point about this, since I'm sure it will come up as this becomes used and the change is trivial in the config file mentioned above. I'm about to import my huge library and will report on whether 255 works for me personally, but that is still a lot of characters to store pretty much anything, even absurdly long titles, hashes, and metadata, so I'm curious what people's opinions of this are. Should we aim for UTF8 compatibility, or do we require the length?

**Comment by joshuaboniface** [comment]
It actually seems like a `latin1` encoding won't work properly:

> org.springframework.jdbc.UncategorizedSQLException: PreparedStatementCallback; uncategorized SQLException for SQL [select id, path, folder, type, format, title, album, artist, album_artist, disc_number, track_number, year, genre, bit_rate, variable_bit_rate, duration_seconds, file_size, width, height, cover_art_path, parent_path, play_count, last_played, comment, created, changed, last_scanned, children_last_updated, present, version from media_file where path=?]; SQL state [HY000]; error code [1267]; (conn:105688) Illegal mix of collations (latin1_general_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation '='

**Comment by joshuaboniface** [comment]
And it seems like I spoke too soon about 255 characters being "more than enough". Spoiled by a C.P.E. Bach piece with a longer path on my system. So it looks like both options are out for me.

I'll do some database tweaking to see if I can find a solution to that.

**Comment by joshuaboniface** [comment]
From my research, I've been able to find out the following things for sure:

1. This is due to the ROW_FORMAT used on the tables. For this particular combination of things (UTF8 + >255 length VARCHARs), we need this option set to "Dynamic", which requires the  InnoDB Barracuda storage engine and large_prefix support (which I already had).
2. MySQL versions before ~5.7 and MariaDB before 10.2 (which is currently Beta) don't have any way to control the ROW_FORMAT on the database/server level and it needs to be set per-table on creation; this is fixed in 5.7/10.2 with the '[innodb_default_row_format](https://mariadb.com/kb/en/mariadb/xtradbinnodb-server-system-variables/#innodb_default_row_format)' variable, which can set the default and I presume avoid this issue; this might be useful if you're on a bleeding-edge version, but not for most users.
3. It seems you can't really work around this by manually creating the tables with the correct values; if you do, Liquidbase just barfs `liquibase.exception.ValidationFailedException` and crashes. It might be possible to create the DB then edit the schema, but I'm not confident enough to try it. This might also affect number 2 above.
4. I researched Liquidbase a bit and it seems we would need to insert a `modifySql` into every table creation section:
    ```
    <modifySql dbms="mysql">
        <append value=" ENGINE=INNODB CHARSET=UTF8 COLLATE utf8_unicode_ci ROW_FORMAT=Dynamic"/>
    </modifySql>
    ```
Given the problems above, I suspect that this is the only real way forward. I've messed around with this a fair bit but I'm having trouble getting it to actually work, mostly because it seems like the `insert` and other elements need to be removed from the creation `changeSet`s or we end up with junky schema like `INSERT INTO libresonic.version (version) VALUES (1) ENGINE=INNODB....`. I'll keep plugging away at this but I doubt I'll get it done soon; if someone knows Liquibase better and can implement this successfully please do!

**Comment by muff1nman** [comment]
Thanks for digging into this.The initial schema just used the max volume size for hsql, which is obviously a problem for MySQL. It would have been ideal to go thru each and every column and decide a proper column size, but I did not as it would be pretty time intensive and error prone.  The next best thing would have been to create a couple of t-shirt sizes if you will. One small varchar, one medium varchar, and one large/max. In the interest of time I decided to just use max everywhere. As you have run into this is still problematic for MySQL. So another way forward would be to go back thru and trim back some columns to use smaller sizes.

**Comment by joshuaboniface** [comment]
Glad to help! I made the original request for external DB since I basically broke the flat file format with a 1.7T library (once I hit a certain point it just wouldn't import new music)! So getting this working is a big deal for me ;-)

As mentioned, at least for me personally, I do have at least a few paths that are longer than 255 characters, so any solution would have to allow both the UTF8 formatting and long values, at least for certain tables, and I suspect I'm not the only user who might face that eventually, especially those with Classical collections that feature ridiculously long artist/song names.

From my research about Liquibase it seems like defining the alternate ROW_FORMAT for MySQL databases in general might be better in the long run, especially since Dynamic tables have been supported for several versions now, and this lets us just go with 512 in general and keep the initial config simpler. Either way we'll be making some pretty extensive changes to the schema.

I'll keep plugging away at it since I think I understand the Liquibase format now, and will report once I have some results!

**Comment by joshuaboniface** [comment]
That actually turned out to be quite easy; I created a `schema1.xml`, which contains only the `<createTable>` (and conditional) sections in a single large `<changeSet>`, with the `<modifySql>` at the end, and it worked perfectly, thanks to the conditionals in the remaining sections. It's a bit of a hack so I'm not going to outright contribute it, but for now it got everything working for me.

**Comment by muff1nman** [comment]
It doesn't sound right that every create table would need to be changed. We should be able to change the default settings.

**Comment by joshuaboniface** [comment]
I agree, it seems ridiculously tedious to change each one, especially with what I found trying to use modifySql. If there's a global option (perhaps set by a flag in the config) that would be far better.

**Comment by joshuaboniface** [comment]
So I have looked more into this, and gotten nowhere, so my personal workaround has been to upgrade to the (now-Release Candidate) MariaDB 10.2 branch, which sets the ROW_FORMAT to `Dynamic` by default and works perfectly with the database schema as-is. I'll close out the issue but if you're using an older database version and have this problem, upgrading is the easiest option.

**Comment by muff1nman** [comment]
Good to know, unfortunately I never got a chance to look into it myself. In fact I think I was on the 10.2 version when I initially tested this and hence why I never ran into the problem. With that said, I agree that anyone looking to use an older MySQL/MariaDB should upgrade or look to change their default row_format to Dynamic.

**Comment by joshuaboniface** [comment]
Fair enough, I think encouraging new DBs is a good policy in any case, since Liquibase seems really limited in fixing this without a big schema rewrite.

The original workaround I used before upgrading was:

1. Set the VarChar size to 255 in the code.
2. Build and install, allowing the system to create the tables with 255 size.
3. Stop tomcat/Libresonic, use `ALTER TABLE` (MySQL script) to increase the VarChar size (I upped it to 1024).
4. Edit the VarChar size to match the size set in 3.
5. Build and install - the database should just work at this point.

I did run into a problem when trying to move to 6.2-beta2, which is what prompted me to look at this again, but I suspect doing it fresh again would have worked using this same procedure. Hopefully this helps someone!
